codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java

```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
+
                 // Provide the user the opportunity to specify the streamId
-                int streamId = getStreamId(httpMsg.headers());
+                currentStreamId = getStreamId(httpMsg.headers());
 
                 // Convert and write the headers.
                 Http2Headers http2Headers = HttpUtil.toHttp2Headers(httpMsg);
-                Http2ConnectionEncoder encoder = encoder();
-
-                if (hasData) {
-                    promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, false, promiseAggregator.newPromise());
-                    httpMsgNeedRelease = false;
-                    encoder.writeData(ctx, streamId, httpMsg.content(), 0, true, promiseAggregator.newPromise());
-                    promiseAggregator.doneAllocatingPromises();
-                } else {
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, true, promise);
-                }
-            } catch (Throwable t) {
-                if (promiseAggregator == null) {
-                    promise.tryFailure(t);
-                } else {
-                    promiseAggregator.setFailure(t);
+                endStream = msg instanceof FullHttpMessage && !((FullHttpMessage) msg).content().isReadable();
+                encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, endStream, promiseAggregator.newPromise());
+            }
+
+            if (msg instanceof HttpContent) {
+                final HttpContent httpContent = (HttpContent) msg;
+                final ByteBuf content = httpContent.content();
+                if (!endStream) {
+                    endStream = msg instanceof LastHttpContent && ((LastHttpContent) msg).trailingHeaders().isEmpty();
+
+                    // Write the data
+                    release = false;
+                    encoder.writeData(ctx, currentStreamId, content, 0, endStream, promiseAggregator.newPromise());
                 }
-            } finally {
-                if (httpMsgNeedRelease) {
-                    httpMsg.release();
+
+                if (!endStream && msg instanceof LastHttpContent) {
+                    // Convert and write any trailing headers.
+                    final LastHttpContent lastContent = (LastHttpContent) httpContent;
+                    final Http2Headers headers = lastContent.trailingHeaders().isEmpty() ? null :
+                            HttpUtil.toHttp2Headers(lastContent.trailingHeaders());
+
+                    if (headers != null && !headers.isEmpty()) {
+                        encoder.writeHeaders(ctx, currentStreamId, headers, 0, true, promiseAggregator.newPromise());
+                    }
                 }
             }
-        } else {
-            ctx.write(msg, promise);
+
+            promiseAggregator.doneAllocatingPromises();
+        } catch (Throwable t) {
+            promiseAggregator.setFailure(t);
```

Comment:
@nmittler - I don't think we should have to call `donAllocatingPromises()` in the event of failure. I think the `SimpleChannelPromiseAggregator` should support termination by calling `setFailure(..)`.